includes/rest/class-server.php

```diff
@@ -0,0 +1,103 @@
+<?php
+namespace Activitypub\Rest;
+
+use WP_REST_Response;
+use Activitypub\Signature;
+
+/**
+ * ActivityPub Server REST-Class
+ *
+ * @author Django Doucet
+ *
+ * @see https://www.w3.org/TR/activitypub/#security-verification
+ */
+class Server {
+
+	/**
+	 * Initialize the class, registering WordPress hooks
+	 */
+	public static function init() {
+		\add_filter( 'rest_request_before_callbacks', array( self::class, 'authorize_activitypub_requests' ), 10, 3 );
+		\add_action( 'rest_api_init', array( self::class, 'register_routes' ) );
+	}
+
+	/**
+	 * Register routes
+	 */
+	public static function register_routes() {
+		\register_rest_route(
+			'activitypub/1.0',
+			'/application',
+			array(
+				array(
+					'methods'             => \WP_REST_Server::READABLE,
+					'callback'            => array( self::class, 'application_actor' ),
+					'permission_callback' => '__return_true',
+				),
+			)
+		);
+	}
+
+	/**
+	 * Render Application actor profile
+	 *
+	 * @return WP_REST_Response
+	 */
+	public static function application_actor() {
+		$json = new \stdClass();
+
+		$json->{'@context'} = \Activitypub\get_context();
+		$json->id = \get_rest_url( null, 'activitypub/1.0/application' );
+		$json->type = 'Application';
+		$json->preferredUsername = wp_parse_url( get_site_url(), PHP_URL_HOST ); // phpcs:ignore WordPress.NamingConventions
+		$json->name = get_bloginfo( 'name' );
+		$json->summary = 'WordPress-ActivityPub application actor';
+		$json->manuallyApprovesFollowers = true; // phpcs:ignore WordPress.NamingConventions
+		$json->icon = array( get_site_icon_url() ); // phpcs:ignore WordPress.NamingConventions short array syntax
+		$json->publicKey = (object) array( // phpcs:ignore WordPress.NamingConventions
+			'id' => \get_rest_url( null, 'activitypub/1.0/application#main-key' ),
+			'owner' => \get_rest_url( null, 'activitypub/1.0/application' ),
+			'publicKeyPem' => Signature::get_public_key( -1 ), // phpcs:ignore WordPress.NamingConventions
+		);
+
+		$response = new WP_REST_Response( $json, 200 );
+
+		$response->header( 'Content-Type', 'application/activity+json' );
+
+		return $response;
+	}
+
+	/**
+	 * Callback function to authorize each api requests
+	 *
+	 * @see \WP_REST_Request
+	 *
+	 * @param                  $response
+	 * @param                  $handler
+	 * @param \WP_REST_Request $request
+	 *
+	 * @return mixed|\WP_Error
+	 */
+	public static function authorize_activitypub_requests( $response, $handler, $request ) {
+		$route = $request->get_route();
+		if ( str_starts_with( $route, '/activitypub' ) ) {
```

Comment:
@mediaformat what do you think about an `verify` filter, where we hook the signature mechansim in, but that could also be implemented by others? Like WordPress is doing with the login for example: `_wp_get_current_user` https://github.com/WordPress/WordPress/blob/f28299d50422286d8b2c021e5405a3b2fec44a22/wp-includes/user.php#L3581

Reply:
That would allow us, to add other verification methods like a simple Webmention/Pingback check if signature is not available: https://www.w3.org/TR/webmention/#request-verification